### PR TITLE
Add disable preview toggle in settings panel

### PR DESCRIPTION
**Pull request by SajidAlamQB**
## Description
Related to: #1921

This feature enables users to control data previews for all nodes from the settings panel of Kedro-Viz when running kedro viz locally.

After running Kedro-Viz locally, dataset previews for all nodes will be enabled by default. However, users will have the ability to enable or disable this feature from the Kedro-Viz settings panel.

## Development Notes

**Backend Changes:**

- Introduced a new API endpoint `/api/preferences` to handle user preferences updates.
- Created a `UserPreference` model to define the preference schema.
- Implemented a new function in `router.py` to update the `showDatasetPreviews` preference and set the class variable accordingly.
- Adjusted the logic in `responses.py` to use the new preference for determining dataset preview settings.

**Frontend Changes:**

- Added a new `showDatasetPreviews` toggle in the Redux state to manage the setting.
- Updated the `SettingsModal` component to include the `showDatasetPreviews` toggle.

**How to Test:**

- Run kedro viz locally via
- Open the Kedro-Viz settings panel.
- Toggle the Dataset Previews setting on and off.
- Verify that the dataset previews for all nodes are enabled/disabled based on the setting.
- Check the network tab to confirm that Preferences API is called and that the preferences are correctly updated on the backend. (Might need to toggle preserve logs to view it as saving in the settings menu refreshes the page and clears network logs)

New dataset preview toggle:

image

New API endpoint:

image

## Checklist

- [x] Read the [contributing](/CONTRIBUTING.md) guidelines
- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added new entries to the `RELEASE.md` file
- [ ] Added tests to cover my changes

### CODE REVIEW

**Comment by Huongg** [comment]
hey @SajidAlamQB what's the best way to test your branch locally? I run `kedro viz run` in demo-project but I guess it won't reflect any changes from the FE

**Comment by SajidAlamQB** [comment]
> hey @SajidAlamQB what's the best way to test your branch locally? I run `kedro viz run` in demo-project but I guess it won't reflect any changes from the FE

Hi, I'm making some changes at the moment, but I've been using `make build` and `make run` to test the changes locally.

**Comment by Huongg** [comment]
> > hey @SajidAlamQB what's the best way to test your branch locally? I run `kedro viz run` in demo-project but I guess it won't reflect any changes from the FE
> 
> Hi, I'm making some changes at the moment, but I've been using `make build` and `make run` to test the changes locally.

oh i see, maybe i got it wrong but from the description it says when you run "kedro viz run". So i expect if i run "kedro viz run" the behaviour will be different from if i run "kedro viz build" etc. Is it the case?

**Comment by SajidAlamQB** [comment]
> oh i see, maybe i got it wrong but from the description it says when you run "kedro viz run". So i expect if i run "kedro viz run" the behaviour will be different from if i run "kedro viz build" etc. Is it the case?

It should be the same behaviour, I've made quite a few changes and updated the description might be worth testing again.

File: src/actions/index.js [hunk 1]
```diff
@@ -141,6 +141,51 @@ export function toggleShowFeatureHints(showFeatureHints) {
   };
 }
 
+export const UPDATE_PREFERENCES = 'UPDATE_PREFERENCES';
```
Comment (rashidakanchwala) [inline src/actions/index.js:144]: I wouldn't add this as a redux action. I would just have a useEffect in settings.modal that would track changes to `setShowDataSetPreview` and make the below call.

File: src/reducers/index.js [hunk 2]
```diff
@@ -81,6 +82,11 @@ const combinedReducer = combineReducers({
     TOGGLE_SHOW_FEATURE_HINTS,
     'showFeatureHints'
   ),
+  showDatasetPreviews: createReducer(
```
Comment (Huongg) [inline src/reducers/index.js:85]: hey if we can have a test here to cover this change, would be great

File: src/components/settings-modal/settings-modal.js [hunk 3]
```diff
@@ -80,19 +90,49 @@ const SettingsModal = ({
...
+        if (!response.ok) {
+          throw new Error('Failed to update preferences');
+        }
+
+        const result = await response.json();
+        console.log('Preferences updated successfully:', result);
+      } catch (error) {
+        console.error('Error updating preferences:', error);
+      }
+    };
```
Comment (jitu5) [inline src/components/settings-modal/settings-modal.js:124]: Please move this api call to separate file.

File: src/components/settings-modal/settings-modal.js [hunk 4]
```diff
@@ -80,19 +90,49 @@ const SettingsModal = ({
...
+
+    updatePreferences();
+  }, [showDatasetPreviewsValue]);
```
Comment (jitu5) [inline src/components/settings-modal/settings-modal.js:127]: I see few problems related to api calls
1. On first load every time it make `api/preferences` api call
2. Whenever user clicks on toggle button it make `api/preferences `api call, ideally it should call only when clicked on save button.
3. When user clicks on toggle button and it saves the preferences even after user clicks on cancel button.

File: src/components/settings-modal/settings-modal.js [hunk 5]
```diff
@@ -65,6 +74,7 @@ const SettingsModal = ({
 
         onToggleIsPrettyName(isPrettyNameValue);
         onToggleShowFeatureHints(showFeatureHintsValue);
+        onToggleShowDatasetPreviews(showDatasetPreviewsValue);
```
Comment (Huongg) [inline src/components/settings-modal/settings-modal.js:77]: do you think we can have the  update state `onToggleShowDatasetPreviews(showDatasetPreviewsValue);` and API call together? 

perhaps something here like 

`updateShowDatasetPreviews()`

```
const updateShowDatasetPreviews = () => {
    // to update redux state
    onToggleShowFeatureHints(showFeatureHintsValue);

   // to update BE
   const updatePreferences = async () => {
      try {
        const response = await fetch('/api/preferences', {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            showDatasetPreviews: showDatasetPreviewsValue,
          }),
        });

        if (!response.ok) {
          throw new Error('Failed to update preferences');
        }

        const result = await response.json();
        console.log('Preferences updated successfully:', result);
      } catch (error) {
        console.error('Error updating preferences:', error);
      }
    };

    updatePreferences();
}
```

It's just easier to know that all the actions related to update dataset preview are in one place

File: package/kedro_viz/api/rest/router.py [hunk 6]
```diff
@@ -49,6 +50,18 @@ async def get_single_node_metadata(node_id: str):
     return get_node_metadata_response(node_id)
 
 
+@router.post("/preferences")
```
Comment (jitu5) [inline package/kedro_viz/api/rest/router.py:53]: Any reason why we created a new api endpoint for setting preferences instead of using existing api/nodes/ by passing user preferences set in redux state to api/nodes/

**Review by Huongg — COMMENTED** [review]
thank you @SajidAlamQB. I've left a few comments. Let me know if you have any questions about them

**Review by stephkaiser — APPROVED** [review]

File: src/components/settings-modal/settings-modal.js (hunk 3 above)
Comment (SajidAlamQB) [reply src/components/settings-modal/settings-modal.js]: Moved it to `src/utils/preferences-api.js`.

File: src/components/settings-modal/settings-modal.js (hunk 5 above)
Comment (SajidAlamQB) [reply src/components/settings-modal/settings-modal.js:92]: The API call moved to a separate file in `src/utils/preferences-api.js`.

File: src/components/settings-modal/settings-modal.js (hunk 4 above)
Comment (SajidAlamQB) [reply src/components/settings-modal/settings-modal.js]: I moved the API call to update preferences inside the `useEffect` that is triggered when the "Save and apply" button is clicked.

File: package/kedro_viz/api/rest/router.py [hunk 7]
```diff
@@ -49,6 +50,18 @@ async def get_single_node_metadata(node_id: str):
...
+async def update_preferences(preferences: UserPreference):
+    try:
+        DataNodeMetadata.set_is_all_previews_enabled(preferences.showDatasetPreviews)
+        return {"message": "Preferences updated successfully"}
+    except Exception as exception:
+        logger.error("Failed to update preferences: %s", str(exception))
+        raise HTTPException(
```
Comment (ravi-kumar-pilla) [inline package/kedro_viz/api/rest/router.py:60]: since we are handling the exception, it would be nice to return a JSONResponse as below -
return JSONResponse(status_code=500, content={"message": f"{exc}"}) if we want the users to see the err trace message on the console. If not we can have a generic message and just log the exception.

File: src/components/settings-modal/settings-modal.js [hunk 8]
```diff
@@ -40,12 +44,31 @@ const SettingsModal = ({
...
     setShowFeatureHintsValue(showFeatureHints);
   }, [showFeatureHints]);
 
+  useEffect(() => {
+    setShowDatasetPreviewsValue(showDatasetPreviews);
+  }, [showDatasetPreviews]);
+
+  useEffect(() => {
+    setToggleFlags(flags);
+  }, [flags]);
```
Comment (ravi-kumar-pilla) [inline src/components/settings-modal/settings-modal.js:61]: I see this effect was missing earlier. Why is this change needed ?

File: package/kedro_viz/api/rest/router.py [hunk 9]
```diff
@@ -49,6 +50,18 @@ async def get_single_node_metadata(node_id: str):
     return get_node_metadata_response(node_id)
 
 
+@router.post("/preferences")
+async def update_preferences(preferences: UserPreference):
+    try:
+        DataNodeMetadata.set_is_all_previews_enabled(preferences.showDatasetPreviews)
+        return {"message": "Preferences updated successfully"}
```
Comment (ravi-kumar-pilla) [inline package/kedro_viz/api/rest/router.py:57]: You need to send the success message as - `return JSONResponse(status_code=200, content={"message": "Preferences updated successfully"})`. If not this will be printed as a message in the browser console.

File: src/store/store.js [hunk 10]
```diff
@@ -57,6 +57,7 @@ const saveStateToLocalStorage = (state) => {
     theme: state.theme,
     isPrettyName: state.isPrettyName,
     showFeatureHints: state.showFeatureHints,
+    showDatasetPreviews: state.showDatasetPreviews,
```
Comment (ravi-kumar-pilla) [inline src/store/store.js:60]: Hi @SajidAlamQB, 

There is an issue setting `showDatasetPreviews` this way. For other flags like `showFeatureHints` we are not sending a request to backend for updating the toggleStatus. So either having the default or reading from localStorage works. 

But for this specific flag, we need to get the initial status from the backend. This would require you to create another api `/preferences` which will be a GET request. This way the FE and backend will be in sync even when the browser is changed or the localStorage is deleted. Let me know if this is not clear. Thank you

Comment (SajidAlamQB) [reply src/store/store.js:60]: Could you elaborate on this a bit more, don't we assume that `showDatasetPreviews` is enabled by default?

Comment (ravi-kumar-pilla) [reply src/store/store.js:60]: Let's connect and discuss

File: src/config.js [hunk 11]
```diff
@@ -77,8 +78,23 @@ export const settings = {
...
 
+// Fetch preferences from the backend and update settings
+(async () => {
+  try {
+    const preferences = await fetchPreferences();
+    settings.showDatasetPreviews.default = preferences.showDatasetPreviews;
+  } catch (error) {
+    console.error('Error fetching initial preferences:', error);
+  }
+})();
```
Comment (jitu5) [inline src/config.js:96]: @SajidAlamQB I think this might not be the right place to make call, as its async, it may or may not update Redux state, I think you should explore redux-thunk.

Comment (ravi-kumar-pilla) [reply src/config.js:96]: @SajidAlamQB, As I was thinking the api call can be delayed and it can be called when someone clicks on the settings icon (before we open the settings modal). That way you can dispatch an action when someone clicks on settings icon -> get the preferences json -> update the toggle accordingly -> open the modal popup which has the updated state. The other way is to call the api along with /main on first load, similar to /api/package-compatibilities. 

This file can be left with the default as before. We will only update redux store.

File: src/components/settings-modal/settings-modal.js [hunk 12]
```diff
@@ -40,12 +46,33 @@ const SettingsModal = ({
...
+  useEffect(() => {
+    if (visible.settingsModal) {
+      getPreferences();
+    }
+  }, [visible.settingsModal, getPreferences]);
+
+  const handleSavePreferences = useCallback(async () => {
+    try {
+      const result = await updatePreferences(showDatasetPreviewsValue);
+      console.log('Preferences updated successfully:', result);
```
Comment (jitu5) [inline src/components/settings-modal/settings-modal.js:70]: Please remove success console.

File: src/actions/index.js [hunk 13]
```diff
@@ -141,6 +141,19 @@ export function toggleShowFeatureHints(showFeatureHints) {
   };
 }
 
+export const TOGGLE_SHOW_DATASET_PREVIEWS = 'TOGGLE_SHOW_DATASET_PREVIEWS';
```
Comment (ravi-kumar-pilla) [inline src/actions/index.js:144]: can we have all the actions related to preferences in 1 file like - `src/actions/preferences.js` as you added preferencesLoaded ?

File: src/components/settings-modal/settings-modal.js [hunk 14]
```diff
@@ -40,12 +46,32 @@ const SettingsModal = ({
...
   useEffect(() => {
     setShowFeatureHintsValue(showFeatureHints);
   }, [showFeatureHints]);
 
+  useEffect(() => {
+    setShowDatasetPreviewsValue(showDatasetPreviews);
+  }, [showDatasetPreviews]);
+
+  useEffect(() => {
+    if (visible.settingsModal) {
```
Comment (ravi-kumar-pilla) [inline src/components/settings-modal/settings-modal.js:62]: is this check required ? Also, I wonder how the toggle changes if the api takes time to return. I felt we can get the state before settingsModal is open. Based on the code, it seems the state is fetched after the modal is visible

File: src/components/settings-modal/settings-modal.js [hunk 15]
```diff
@@ -209,13 +251,17 @@ export const mapStateToProps = (state) => ({
   flags: state.flags,
   showFeatureHints: state.showFeatureHints,
   isPrettyName: state.isPrettyName,
+  showDatasetPreviews: state.preferences.showDatasetPreviews,
```
Comment (jitu5) [inline src/components/settings-modal/settings-modal.js:254]: It should be like this right?
`showDatasetPreviews: state.showDatasetPreviews,` 
or I am missing something ?

File: src/components/settings-modal/settings-modal.js (hunk 14 above)
Comment (jitu5) [reply src/components/settings-modal/settings-modal.js:62]: @ravi-kumar-pilla Yes `if (visible.settingsModal)` this check will make sure that we get preferences when we open it and not when we close. Current setup is fine as `visible.settingsModal` get true it make api call, and therefore `showDatasetPreviews` changes and this triggers `setShowDatasetPreviewsValue` to update. 

```
  useEffect(() => {
    setShowDatasetPreviewsValue(showDatasetPreviews);
  }, [showDatasetPreviews]);
```

```
  useEffect(() => {
    if (visible.settingsModal) {
      getPreferences();
    }
  }, [visible.settingsModal, getPreferences]);
```

File: src/components/settings-modal/settings-modal.js (hunk 15 above)
Comment (ravi-kumar-pilla) [reply src/components/settings-modal/settings-modal.js:254]: I tested locally and surprisingly it works 🗡️  .... not sure if the prop is used to set the local state of the toggle

Comment (SajidAlamQB) [reply src/components/settings-modal/settings-modal.js:254]: This just keeps `showDatasetPreviews` nested within the preferences state its just to keep it more distinct incase in the future we want to add more preferences. We can also go with the original global state version as well.

File: src/store/initial-state.js [hunk 16]
```diff
@@ -24,6 +24,7 @@ export const createInitialState = () => ({
   expandAllPipelines: false,
   isPrettyName: settings.isPrettyName.default,
   showFeatureHints: settings.showFeatureHints.default,
+  showDatasetPreviews: settings.showDatasetPreviews.default,
```
Comment (ravi-kumar-pilla) [inline src/store/initial-state.js:27]: It is good to have a nesting, but how are you able to nest if the initial state is at the same level as flags ?
https://github.com/kedro-org/kedro-viz/pull/1977/files#r1688315574

File: src/components/settings-modal/settings-modal.js (hunk 15 above)
Comment (ravi-kumar-pilla) [reply src/components/settings-modal/settings-modal.js:254]: it is good to have a nesting, but I guess the question is how showDatasetPreviews is under preferences when the initial state does not have a nesting

File: src/store/initial-state.js (hunk 16 above)
Comment (SajidAlamQB) [reply src/store/initial-state.js:27]: Fair you are right, we can go back to original.

File: src/reducers/index.js [hunk 17]
```diff
@@ -96,6 +100,16 @@ const combinedReducer = combineReducers({
     TOGGLE_SHOW_FEATURE_HINTS,
     'showFeatureHints'
   ),
+  showDatasetPreviews: (state = true, action) => {
```
Comment (ravi-kumar-pilla) [inline src/reducers/index.js:103]: can we have the block inside its own reducer as before ? This is more of a reducer block and best kept separate, instead of keeping it inside combineReducers to be consistent. Let me know if this was needed or if I am missing something here.

File: src/actions/preferences.js [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+// Action Creators
+export const preferencesLoaded = (preferences) => ({
+  type: PREFERENCES_LOADED,
+  payload: preferences,
+});
+
+export const toggleShowDatasetPreviews = (showDatasetPreviews) => ({
+  type: TOGGLE_SHOW_DATASET_PREVIEWS,
+  showDatasetPreviews,
+});
```
Comment (jitu5) [inline src/actions/preferences.js:16]: @SajidAlamQB 
Can we combine `preferencesLoaded` and `toggleShowDatasetPreviews` to have single action like `updateDatasetPreviews`/ `UPDATE_DATASETS_PREVIEWS` as both are updating `showDatasetPreviews` in Redux state.

**Review by jitu5 — APPROVED** [review]
LGTM! Thanks @SajidAlamQB

File: src/actions/preferences.js [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+// Action Creators
+export const updateDatasetPreviews = (showDatasetPreviews) => ({
+  type: UPDATE_DATASET_PREVIEWS,
+  payload: { showDatasetPreviews },
+});
+
+export const getPreferences = () => async (dispatch) => {
+  try {
+    const preferences = await fetchPreferences();
+    dispatch(updateDatasetPreviews(preferences.showDatasetPreviews));
```
Comment (ravi-kumar-pilla) [inline src/actions/preferences.js:15]:
```suggestion
    dispatch(updateUserPreferences(preferences));
```

File: src/reducers/index.js [hunk 20]
```diff
@@ -81,6 +82,7 @@ const combinedReducer = combineReducers({
   modularPipeline,
   visible,
   runsMetadata,
+  showDatasetPreviews,
```
Comment (ravi-kumar-pilla) [inline src/reducers/index.js:85]:
```suggestion
  userPreferences
```

File: src/reducers/preferences.js [hunk 21]
```diff
@@ -0,0 +1,14 @@
+import { UPDATE_DATASET_PREVIEWS } from '../actions/preferences';
+
+const initialState = true;
+
+const showDatasetPreviews = (state = initialState, action) => {
+  switch (action.type) {
+    case UPDATE_DATASET_PREVIEWS:
```
Comment (ravi-kumar-pilla) [inline src/reducers/preferences.js:7]:
```suggestion
    case UPDATE_USER_PREFERENCES:
```

File: src/actions/preferences.js [hunk 22]
```diff
@@ -0,0 +1,19 @@
+import { fetchPreferences } from '../utils/preferences-api';
+
+// Action Types
+export const UPDATE_USER_PREFERENCES = 'UPDATE_USER_PREFERENCES';
+
+// Action Creators
+export const updateUserPreferences = (preferences) => ({
+  type: UPDATE_USER_PREFERENCES,
+  payload: { preferences },
```
Comment (ravi-kumar-pilla) [inline src/actions/preferences.js:9]:
```suggestion
  payload: preferences,
```

Comment (ravi-kumar-pilla) [reply src/actions/preferences.js:9]: preferences should already be an object. Do we need to wrap it around { } ?

File: src/components/settings-modal/settings-modal.test.js [hunk 23]
```diff
@@ -73,5 +74,11 @@ describe('SettingsModal', () => {
       type: 'TOGGLE_IS_PRETTY_NAME',
       isPrettyName: false,
     });
+
+    mapDispatchToProps(dispatch).onToggleShowDatasetPreviews(false);
+    expect(dispatch.mock.calls[3][0]).toEqual({
+      type: 'UPDATE_USER_PREFERENCES',
+      payload: { preferences: { showDatasetPreviews: false } },
```
Comment (ravi-kumar-pilla) [inline src/components/settings-modal/settings-modal.test.js:81]: UPDATE_USER_PREFERENCES should take an object preferences. The key preferences is not needed. The object is a preferences object as a whole

```
{ 
  showDatasetPreviews: false 
}
```

File: src/reducers/preferences.js [hunk 24]
```diff
@@ -0,0 +1,19 @@
...
+const initialState = {
+  showDatasetPreviews: true,
+};
+
+const userPreferences = (state = initialState, action) => {
+  switch (action.type) {
+    case UPDATE_USER_PREFERENCES:
+      return {
+        ...state,
+        showDatasetPreviews: action.payload.preferences.showDatasetPreviews,
```
Comment (ravi-kumar-pilla) [inline src/reducers/preferences.js:12]:
```suggestion
        showDatasetPreviews: action.payload.showDatasetPreviews,
```

Comment (ravi-kumar-pilla) [reply src/reducers/preferences.js:12]: The payload for userPreferences should be a preferences object. i.e., 
```
{
     showDatasetPreviews: false
}
```
It need not contain the preferences key inside like -
```
{
     preferences: {
           showDatasetPreviews: false
      }
}
```
I think there is some confusion around nesting. If we have payload as - `payload = showDatasetPreviews` (this means no nesting and cannot be extended in future)

`payload = {showDatasetPreviews: false}` , there is a nested prop and the payload becomes an object. Here, the object will have fields of User Preferences.

File: src/store/initial-state.js [hunk 25]
```diff
@@ -24,6 +24,9 @@ export const createInitialState = () => ({
   expandAllPipelines: false,
   isPrettyName: settings.isPrettyName.default,
   showFeatureHints: settings.showFeatureHints.default,
+  userPreferences: {
```
Comment (ravi-kumar-pilla) [inline src/store/initial-state.js:27]: This is perfect 💯

**Review by ravi-kumar-pilla — APPROVED** [review]
Thank you for addressing all the PR comments @SajidAlamQB . LGTM and works well 💯
